### PR TITLE
Use UnionCollection to avoid large set allocations

### PR DESCRIPTION
**Pull request by sharwell**

### CODE REVIEW

File: src/Compilers/Core/Portable/Collections/UnionCollection.cs [hunk 1]
```diff
@@ -42,7 +42,24 @@ public static ICollection<T> Create(ICollection<T> coll1, ICollection<T> coll2)
...
+            {
+                case 0:
+                    return SpecializedCollections.EmptyCollection<T>();
+
+                case 1:
+                    return collections[0];
+
+                default:
+                    return new UnionCollection<T>(Flatten(collections));
+            }
```
Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/Collections/UnionCollection.cs:62]: nit: 

```c#
=> collections.Length switch
{
    0 => ...
}
```

File: src/Compilers/Core/Portable/Collections/UnionCollection.cs [hunk 2]
```diff
@@ -68,6 +85,38 @@ private UnionCollection(ImmutableArray<ICollection<T>> collections)
...
+            var builder = ArrayBuilder<ICollection<T>>.GetInstance();
+            foreach (var collection in collections)
+            {
+                if (collection is UnionCollection<T> unionCollection)
+                {
+                    builder.AddRange(unionCollection._collections);
+                }
+                else
+                {
+                    builder.Add(collection);
```
Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/Collections/UnionCollection.cs:102]: potentially ignore empty collections (like what is done above)?

**Review by CyrusNajmabadi — APPROVED** [review]

File: src/Compilers/Core/Portable/Collections/UnionCollection.cs (hunk 1 above)
Comment (sharwell) [reply src/Compilers/Core/Portable/Collections/UnionCollection.cs:62]: Not sure what the compiler style here is.

File: src/Compilers/Core/Portable/Collections/UnionCollection.cs (hunk 2 above)
Comment (sharwell) [reply src/Compilers/Core/Portable/Collections/UnionCollection.cs:102]: At first I was thinking these are read-only but not necessarily immutable, but with the code above it does seem like a reasonable optimization.

File: src/Compilers/Core/Portable/Collections/UnionCollection.cs (hunk 1 above)
Comment (CyrusNajmabadi) [reply src/Compilers/Core/Portable/Collections/UnionCollection.cs:62]: they def use switch expressions.  but nbd.

File: src/Compilers/Core/Portable/Collections/UnionCollection.cs (hunk 2 above)
Comment (CyrusNajmabadi) [reply src/Compilers/Core/Portable/Collections/UnionCollection.cs:102]: actually, my guess is that it's not.  in practice these will never be empty.  :)

**Pull request merged**